internal/pkg/pkg_test.go

```diff
@@ -307,3 +310,204 @@ func TestSubpackages(t *testing.T) {
 		})
 	}
 }
+
+//nolint:scopelint
+func TestSubpackages(t *testing.T) {
+	type variants struct {
+		matcher   []SubpackageMatcher
+		recursive []bool
+		expected  []string
+	}
+
+	testCases := map[string]struct {
+		pkg   *pkgbuilder.RootPkg
+		cases []variants
+	}{
+		"remote and local nested subpackages": {
+			// root
+			//  ├── remote-sub1 (remote)
+			//  │   ├── Kptfile
+			//  │   └── directory
+			//  │       └── remote-sub3 (remote)
+			//  │           └── Kptfile
+			//  └── local-sub1 (local)
+			//      ├── Kptfile
+			//      ├── directory
+			//      │   └── remote-sub3 (remote)
+			//      │       └── Kptfile
+			//      └── local-sub2 (local)
+			//          └── Kptfile
+			pkg: pkgbuilder.NewRootPkg().
```

Comment:
Thanks for adding the structure. It is clear now. 

On a separate note, there was a discussion on other functions PR to create tests with data so that anyone can add more tests without needing to learn about pkgbuilder. https://github.com/GoogleContainerTools/kpt/tree/next/e2e/testdata/fn-render/basicpipeline
We should do that eventually in future.